ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/layered/LayeredPendingTransactions.java

```diff
@@ -95,66 +95,116 @@ public synchronized TransactionAddedResult addLocalTransaction(
 
     final TransactionAddedResult nonceChecksResult =
-        nonceChecks(pendingTransaction, senderNonce, nonceDistance);
+        nonceChecks(pendingTransaction, stateSenderNonce, nonceDistance);
     if (nonceChecksResult != null) {
       return nonceChecksResult;
     }
 
     try {
-      TransactionAddedResult result =
-          prioritizedTransactions.add(pendingTransaction, (int) nonceDistance);
-
-      if (result.equals(REORG_SENDER)) {
-        result = reorgSenderOf(pendingTransaction, (int) nonceDistance);
-      }
-
-      return result;
+      return prioritizedTransactions.add(pendingTransaction, (int) nonceDistance);
     } catch (final Throwable throwable) {
-      // in case something unexpected happened, log this sender txs and force a reorg of his txs
+      return reconcileAndRetryAdd(
+          pendingTransaction, stateSenderNonce, (int) nonceDistance, throwable);
+    }
+  }
+
+  private TransactionAddedResult reconcileAndRetryAdd(
+      final PendingTransaction pendingTransaction,
+      final long stateSenderNonce,
+      final int nonceDistance,
+      final Throwable throwable) {
+    // in case something unexpected happened, log this sender txs and force a reorg of his txs
+    LOG.warn(
+        "Unexpected error {} when adding transaction {}, current sender status {}",
+        throwable,
+        pendingTransaction.toTraceLog(),
+        prioritizedTransactions.logSender(pendingTransaction.getSender()));
+    LOG.warn("Stack trace", throwable);
+    reconcileSender(pendingTransaction.getSender(), stateSenderNonce);
+    try {
+      return prioritizedTransactions.add(pendingTransaction, nonceDistance);
+    } catch (final Throwable throwable2) {
       LOG.warn(
           "Unexpected error {} when adding transaction {}, current sender status {}",
           throwable,
           pendingTransaction.toTraceLog(),
           prioritizedTransactions.logSender(pendingTransaction.getSender()));
       LOG.warn("Stack trace", throwable);
-      reorgSenderOf(pendingTransaction, (int) nonceDistance);
       return INTERNAL_ERROR;
     }
   }
 
-  private TransactionAddedResult reorgSenderOf(
-      final PendingTransaction pendingTransaction, final int nonceDistance) {
-    final var existingSenderTxs = prioritizedTransactions.getAllFor(pendingTransaction.getSender());
+  private boolean nonceSkewDetected(
+      final PendingTransaction pendingTransaction, final long stateSenderNonce) {
+    final OptionalLong maybeTxPoolSenderNonce =
+        prioritizedTransactions.getCurrentNonceFor(pendingTransaction.getSender());
+    if (maybeTxPoolSenderNonce.isPresent()) {
+      final long txPoolSenderNonce = maybeTxPoolSenderNonce.getAsLong();
+      if (stateSenderNonce != txPoolSenderNonce) {
+        LOG.atDebug()
+            .setMessage(
+                "Nonce skew detected when adding pending transaction {}. Account nonce from world state is {} while current txpool nonce is {}")
```

Comment:
elsewhere are we calling this "nonce gap"?

Reply:
this is a different scenario than "nonce gap", since "nonce gap" is the difference between the account nonce and the pending transaction, while "nonce skew" is the difference between the account nonce from the world state, and the account nonce as seen from the txpool point of view. Will add some doc to explain this better.